### PR TITLE
feat: add browser tools

### PR DESCRIPTION
**Pull request by njhale**
Add MVP `Browser` tools with updated session management for use in Otto.

This means:
- The `Browser` tool daemon doesn't store session data in the workspace directory
-- which is now deleted after each tool call -- and instead creates an independent session directory and browser process for each unique workspace ID it encounters 
- The session directories and browser processes are still ephemeral, but now have a 10 minute TTL from last use
- Workspace IDs are taken from the headers of requests to the tool daemon (can't use the environment variable anymore because tool daemons are effectively singletons in Otto)

### CODE REVIEW

File: browser/tool.gpt [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+---
+Name: Forward
+Share Context: Browser Context
+Tools: service
+Description: Navigates to forward to the next site in history.
+Params: tabID: (required) The ID of the tab.
+Tools: service
+
+#!http://service.daemon.gptscript.local/forward
```
Comment (njhale) [inline browser/tool.gpt:90]: A lot of the tools here aren't useful on their own unless they share pretty much all of the other tools.

What would folks think if I tag everything but the bundle and `Get Page Content` with `Metadata: index: false`?

File: browser/src/session.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+export class SessionManager {
+  private readonly sessions = new Map<string, ManagedSession>()
+  private readonly sessionsLock: Mutex = new Mutex()
+
+  private constructor () {
+  }
+
+  static async create (): Promise<SessionManager> {
+    sessionManager ??= new SessionManager()
+    return sessionManager
```
Comment (njhale) [inline browser/src/session.ts:40]: Session directories can become orphaned if a daemon with "living" sessions is killed. Should we delete orphaned directories at when the session manager is created to handle that?

File: browser/tool.gpt (hunk 1 above)
Comment (thedadams) [reply browser/tool.gpt:90]: I think that makes sense.

**Review by thedadams — APPROVED** [review]

File: browser/src/session.ts (hunk 2 above)
Comment (g-linville) [reply browser/src/session.ts:40]: I think this is a good idea. Orphaned session directories could end up taking up a lot of drive space.

**Review by g-linville — APPROVED** [review]

File: browser/src/server.ts [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+import { randomBytes } from 'node:crypto'
+import { getSessionId, SessionManager } from './session.ts'
+
+async function main (): Promise<void> {
+  console.log('Starting browser server')
+
+  const app = express()
+  // Get port from the environment variable or use 9888 if it is not defined
+  const port = process.env.PORT ?? 9888
+  delete (process.env.GPTSCRIPT_INPUT)
```
[the remark this answers is not in the document]
Comment (g-linville) [reply browser/src/server.ts:17]: I think I may have been the one to add this in the old browser tool repo. I was seeing cases where GPTSCRIPT_INPUT would be passed to internal tool calls and possibly mess them up, if they were trying to read their own args from that env var. At least, I think that's why I was deleting this here.

File: browser/src/session.ts (hunk 2 above)
Comment (njhale) [reply browser/src/session.ts:52]: done

File: browser/src/server.ts [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        } else {
+          page = await browserContext.newPage()
+          openPages.set(tabID, page)
+        }
+        await page.bringToFront()
+
+        let allElements = false
+        if (data.allElements === 'true' || data.allElements === true) {
+          allElements = true
+        }
```
[the remark this answers is not in the document]
Comment (njhale) [reply browser/src/server.ts]: `allElements` wasn't actually being used anymore. Removed.

File: browser/tool.gpt (hunk 1 above)
Comment (njhale) [reply browser/tool.gpt]: done

**Pull request merged**